thoth/storages/graph/postgres.py

```diff
@@ -3775,6 +3775,82 @@ def create_python_cve_record(
 
                 return False
 
+    def update_missing_flag_package_version(
+        self,
+        package_name: str,
+        package_version: str,
+        index_url: str,
+        value: bool,
+    ) -> None:
+        """Update value of is_missing flag for PythonPackageVersion."""
+        with self._session_scope() as session:
+            subq = (session.query(PythonPackageVersion)
+                    .join(PythonPackageIndex)
+                    .filter(PythonPackageVersion.package_name == package_name)
+                    .filter(PythonPackageVersion.package_version == package_version)
+                    .filter(PythonPackageIndex.url == index_url)
+                    .with_entities(PythonPackageVersion.id)
+                    )
+            (
+                session.query(PythonPackageVersion)
+                .filter(PythonPackageVersion.id.in_(subq))
+                .update({"is_missing": value}, synchronize_session='fetch')
+            )
+
+    def get_all_repositories_using_python_package(
+        self,
+        package_name: str,
+        index_url: str,
+        package_version: str = None,
+    ) -> List[str]:
+        """Given package version info, get unique urls of repos which have been advised with this package version."""
+        with self._session_scope() as session:
+            result = (
+                session.query(PythonPackageVersion)
+                .join(PythonPackageIndex)
+                .join(PythonRequirementsLock)
+                .join(AdviserRun, AdviserRun.user_software_stack_id == PythonRequirementsLock.python_software_stack_id)
+                .order_by(AdviserRun.origin)
+                .order_by(AdviserRun.datetime.desc())
+                .distinct(AdviserRun.origin)
+                .filter(PythonPackageIndex.url == index_url)
+                .filter(PythonPackageVersion.package_name == package_name)
+                .filter(PythonPackageVersion.package_version == package_version)
+            )
+            if package_version is not None:
+                result.filter(PythonPackageVersion.package_version == package_version)
+
+            result = result.with_entities(AdviserRun.origin).all()
+            result = [x[0] for x in result]
+            return result
```

Comment:
In any case to be coherent with conventions and queries in general: - `get_adviser_run_origins_all` should retrieve all origins in adviser run, then filters can be applied: 

- `package_name`, `package_version`, `index_url`, `distinct` also should be there because if we want to count how many for a specific origin we can reuse logic across queries. 

- `start_offset` and `count` should be considered if we want to consider pagination purposes.

Moreover `package_name` and `package_version` should be used with normalize method:

`package_name = self.normalize_python_package_name(package_name)`
`package_version = self.normalize_python_package_version(package_version)`

What about?

```
    def get_adviser_run_origins_all(
        self,
        *,
        package_name: str = None,
        package_version: str = None,
        index_url: str = None,
        start_offset: int = 0,
        count: Optional[int] = DEFAULT_COUNT,
        distinct: bool = False,
    ) -> List[str]:
        """Retrieve all origins (git repos URLs) in Adviser Run.

        Examples:
        >>> from thoth.storages import GraphDatabase
        >>> graph = GraphDatabase()
        >>> graph.get_adviser_run_origins_all()
        [HERE SHOULD GO AN EXAMPLE OF A RESULT]
        """
        with self._session_scope() as session:

            query = (
                session.query(AdviserRun)
                .order_by(AdviserRun.origin)
                .order_by(AdviserRun.datetime.desc())
            )

            if package_version or package_name:
                query = (
                    query.join(PythonRequirementsLock, PythonRequirementsLock.python_software_stack_id == AdviserRun.user_software_stack_id)
                    .join(PythonPackageVersion)
                )

                if index_url:
                    query = query.join(PythonPackageIndex)

            query = query.with_entities(AdviserRun.origin)

            if package_name is not None:
                package_name = self.normalize_python_package_name(package_name)
                query.filter(PythonPackageVersion.package_name == package_name)

            if package_version is not None:
                package_version = self.normalize_python_package_version(package_version)
                query.filter(PythonPackageVersion.package_version == package_version)

            if index_url is not None:
                query.filter(PythonPackageIndex.url  == index_url)

            query = query.offset(start_offset).limit(count)

            if distinct:
                query = query.distinct()

            result = [r[0] for r in query.all() if r[0]] # We do not consider None results

            return result
```

Reply:
@fridex PTAL :)